File: 637.Average of Levels in Binary Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var averageOfLevels = function(root) {
    let result = [];
    let stack = [root];
    let count = 1;
    let current = 0;
    let temp = 0;
    while(stack.length) {
        const n = stack.shift();
        if (n) {
            current++;
            temp += n.val;
            if (n.left) {
                stack.push(n.left);
            }
            if (n.right) {
                stack.push(n.right);
            }
            if (current === count) {
                result.push(temp / count);
                temp = 0;
                count = stack.length;
                current = 0;
            }
        }
    }
    return result;
};

File: Golang/Tree/637.average-of-levels-in-binary-tree.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func averageOfLevels(root *TreeNode) []float64 {
	ret := []float64{}
	if root != nil {
		stack := []*TreeNode{root}
		for len(stack) > 0 {
			children := []*TreeNode{}
			sum := 0
			for i := 0; i < len(stack); i++ {
				node := stack[i]
				sum += node.Val
				if node.Left != nil {
					children = append(children, node.Left)
				}
				if node.Right != nil {
					children = append(children, node.Right)
				}
			}
			ret = append(ret, float64(sum)/float64(len(stack)))
			stack = children
		}
	}
	return ret
}
